Enable & customize bulk edit, updates to Billing Module, and other bug fixes/updates.

labkey-bpatel opened the pull request:
- Ticket 38766: Charge with Animal IDs, Batch, Bulk Edit - enable & customize bulk edit feature for both data entry forms.
- Ticket 39285: Changes to the billing module : create group-category associations, update column names and have data under appropriate sounding column names.
- Ticket 39219: Cleanup & Refactor on columns associated with Chargeable Items
- Ticket 39134: Problem with adding charges without animal id : remove isActive filtering
- Ticket 38953: duplicate accounts seen in test server : tierRate column in aliases now FKs to activeTierRates query.
- Ticket 38909: Filter Projects based on Investigator selection for Billing Data Access
- Ticket 39107: Billing Period Summary By Project : This report is now upgraded and renamed to 'View Billing Queries' report to allow users/testers to view billing queries that get processed during a billing run.

labkey-bpatel committed:
Filter accounts based on active tierRates - fix for these tickets:
-Ticket 38953: duplicate accounts seen in test server and
-Ticket 39134: Problem with adding charges without animal id

labkey-bpatel committed:
Redo 'Estimated Charges By Project' to give a better/fuller view of the billing period, with filtering based on query and project or account. Report is now named 'View Billing Queries'.

labkey-bpatel committed:
Remove filter from the default view.

labkey-bpatel committed:
Ticket 38909: Filter Projects based on Investigator selection for Billing Data Access
- Customized insert and update page for ehr_billing.dataAccess (wip).

labkey-bpatel committed:
Ticket 38909: Filter Projects based on Investigator selection for Billing Data Access
- Conditionalize project dropdown upon investigator selection.

labkey-bpatel committed:
Handle case when project is no longer associated with the investigator.

labkey-bpatel committed:
Don't allow udpates to invoicedItems. Update to Investigator facing query.

labkey-bpatel committed:
Minor corrections to the Investigator facing sub-query.

labkey-bpatel committed:
Bulk Edit related customizations (wip).

labkey-bpatel committed:
Set queryMode.

labkey-bpatel committed:
Bulk Edit Customizations.

labkey-bpatel committed:
Open customized bulk edit window from Add Batch window.

labkey-bpatel committed:
Set queryMode.

labkey-bpatel committed:
Enable customized Bulk Edit for Non-Animal Charges Data Entry (wip).

labkey-bpatel committed:
Update store loading to avoid lag.

labkey-bpatel committed:
Remove additional filter.

labkey-bpatel committed:
Enable customized Bulk Edit for Non-Animal Charges Data Entry.

labkey-bpatel committed:
Enable customized Bulk Edit for Non-Animal Charges Data Entry.

labkey-bpatel committed:
Remove serviceCode references from the view.

labkey-bpatel committed:
Use newly added chargeGroup in MiscCharges table (instead of chargetype, which is now re-purposed for 'Adjustment').

labkey-bpatel committed:
Ticket 39285: Changes to the billing module
- Create new charge group & charge category association table, and add link to finance page.
- Move existing miscCharges data into appropriate sounding column to avoid confusion.
- Update chargeableItems query.xml to use wnprc_billing.chargeCategoryAssociations for Group & Category columns.
- Add validation during bulk rates upload -- ensure group and category combination are from wnprc_billing.chargeCategoryAssociations table.

labkey-bpatel committed:
Remove custom Bulk Edit files (customization done in *Field.js files)

labkey-bpatel committed:
Additional queries to show investigator via investigatorId from ehr.project and ehr_billing.alias tables.

labkey-bpatel committed:
Add cols to metadata file, and useColumnOrder to show fields in that order on data entry forms.

labkey-bpatel committed:
Data entry field customizations for bulk edit and grid panel.

labkey-bpatel committed:
Upgrade version. Remove sort.

labkey-bpatel committed:
Remove unnecessary store creation.

labkey-bpatel committed:
Add chargeGroup to the query

labkey-bpatel committed:
Add createdby to miscChargesFeeRates view.

labkey-bpatel committed:
Pass through performed by for Blood Draws view.

labkey-bpatel committed:
Group Category Associations -- add trigger script, and template so that bulk data can be uploaded using category names (instead of having to map the row id).

labkey-bpatel committed:
Automated test update (wip). Update/add sample data(wip).

labkey-bpatel committed:
Category related updates - additional validation. Sample data & automated test.

labkey-bpatel committed:
Automated test - populate ehr.investigators table, and add investigator from this table into ehr_billing.aliases and ehr.projects.

labkey-bpatel committed:
Automated test - modify provideBillingAccess to insert data via data region with additional check. Move to do this earlier in the test.

labkey-bpatel committed:
Update column source (given that now chargeCategoryId is associated with wnprc_billing.groupCategoryAssociations table)

labkey-bpatel committed:
Automated test - update data entry to use new/different columns.

labkey-bpatel committed:
Fix table name

labkey-bpatel committed:
Revert fk lookup change.

labkey-bpatel committed:
Revert prev. change.

labkey-bpatel committed:
Automated test - updated to use View Billing Queries feature. Additional metadata for the feeRates queries.

labkey-bpatel committed:
Automated test - change from chargetype to groupName for Misc Charges. Update Misc Charges views.

labkey-bpatel committed:
De-couple billing data load from the test.

labkey-bpatel committed:
Automated test - Add Charges With AnimalIds via Bulk Edit.

labkey-bpatel committed:
Automated test - Add Non-Animal Charges via Bulk Edit. Other minor updates.

labkey-bpatel committed:
Automated test - wait on elements.

labkey-bpatel committed:
Set date on render if its not filled in.

labkey-bpatel committed:
Automated test - updates to keep teamcity happy. Additional checks.

labkey-jeckels commented on WNPRC_EHR/test/src/org/labkey/test/tests/wnprc_ehr/WNPRC_EHRTest.java:
```diff
@@ -520,70 +703,109 @@ private void addInvestigators() throws IOException, CommandException
-        assertEquals("One row should be displayed", perDiemFeeRatesTable.getDataRowCount(), 1);
+        assertEquals("One row should be displayed", miscChargesFeeRatesTable.getDataRowCount(), 1);
 
+        DataRegionTable perDiemFeeRatesTable = DataRegionTable.DataRegion(getDriver()).withName("billing-qwp-2").waitFor();
```
These are probably OK to leave as-is since we're not doing them in a for loop, but remember that we can do API-based validation of query results. It's faster and more reliable.

labkey-jeckels commented on wnprc_billing/resources/queries/ehr/aliasesWithInvestigators.sql:
```diff
@@ -0,0 +1,12 @@
+          THEN (investigatorId.lastName ||', '|| investigatorId.firstName)
+          WHEN investigatorId.lastName IS NOT NULL AND investigatorId.firstName IS NULL
+          THEN investigatorId.lastName
+          WHEN investigatorId.lastName IS NULL AND investigatorId.firstName IS NULL
```
Make this an ELSE instead, in case firstName has a value but last name is null

labkey-bpatel replied:
Please review latest commit related to this.

labkey-jeckels commented on wnprc_billing/resources/queries/ehr/project.query.xml:
```diff
@@ -9,6 +9,7 @@
                             <fkDbSchema>ehr</fkDbSchema>
                             <fkTable>investigators</fkTable>
                             <fkColumnName>rowid</fkColumnName>
+                            <fkDisplayColumnName>lastName</fkDisplayColumnName>
```
Is this OK, or do we need to inject a calculated column that does magic with lastName, firstName, and displayName like we're doing in aliasesWithInvestigators?

labkey-jeckels replied:
Yes, injecting a magic calculated column into Investigators seems better so we can use it both here and in aliasesWithInvestigators and projectsWithInvestigators.

labkey-bpatel replied:
Did this by having a new investigatorsWithName query instead of doing it in a customizer.

labkey-jeckels commented on wnprc_billing/resources/queries/ehrLinked/investigators.query.xml:
```diff
@@ -0,0 +1,20 @@
+        <tables xmlns="http://labkey.org/data/xml">
+            <table tableName="investigators" tableDbType="TABLE">
+                <tableTitle>Investigators</tableTitle>
+                <titleColumn>lastName</titleColumn>
```
This could point to the proposed magic calculated column

labkey-bpatel replied:
Removed titleColumn since no longer will be using lastName.

labkey-jeckels commented on wnprc_billing/resources/queries/ehrLinked/project.query.xml:
```diff
@@ -8,6 +8,7 @@
                             <fkDbSchema>ehrLinked</fkDbSchema>
                             <fkTable>investigators</fkTable>
                             <fkColumnName>rowid</fkColumnName>
+                            <fkDisplayColumnName>lastName</fkDisplayColumnName>
```
Do we need this in addition to the titleColumn in investigators.query.xml

labkey-bpatel replied:
this now points to a new investigatorsWithName query.

labkey-jeckels commented on wnprc_billing/resources/queries/ehr_billing/miscCharges/Charges Not Billed.qview.xml:
```diff
@@ -4,8 +4,17 @@
-        <column name="chargecategory"/>
+        <column name="chargeId/name">
+            <properties>
+                <property name="columnTitle" value="Charge Item"/>
```
I notice we're setting this in a few .qview.xml files. Could we apply it centrally in a WNPRC .query.xml file so it catches all the ways it could be displayed instead?

labkey-bpatel replied:
Good point, done!

labkey-jeckels commented on wnprc_billing/resources/queries/wnprc_billing/activeTierRates.sql:
```diff
@@ -0,0 +1,8 @@
+    endDate,
+    isActive
+FROM wnprc_billing.tierRates
+WHERE isActive = 'Y'
```
I'm a little surprised this works, though it's nice. Might be more robust to make this true instead of 'Y'

labkey-bpatel replied:
Done!

labkey-jeckels commented on wnprc_billing/resources/queries/wnprc_billing/groupCategoryAssocWithCategoryName.query.xml:
```diff
@@ -0,0 +1,17 @@
+                <auditLogging>DETAILED</auditLogging>
+                <columns>
+                    <column columnName="chargeGroupName">
+                        <columnTitle>chargeGroupName</columnTitle>
```
Do we really want this to be "chargeGroupName"? And not "Charge Group Name" or something a little more user-oriented?

labkey-bpatel replied:
User-Friendly title now set (with required refactor in the trigger script).

labkey-jeckels reviewed (changes requested):
Overall looks great. A few small questions and suggestions.

labkey-bpatel committed:
Code Review changes & refactors/cleanups:
- Set column title in target query.xml files instead .qview files, esp. for fk columns.
- activeTierRates query update -- Filter on true instead of 'Y'
- investigator field related changes -- create a new query with to derive investigator name, use that instead of lastName. Update billing queries & Data Access customized edit page to use this derived name.
- Update automated test.

labkey-bpatel committed:
Code Review Change - User-friendly titles in template, and related refactor.

labkey-bpatel committed:
Remove titleColumn. Updated to use calculated column for investigator.

labkey-bpatel committed:
Automated test update.

labkey-bpatel committed:
Comment out View Billing Queries test for now (so that rest of the tests to run to completion).

labkey-bpatel committed:
Automated test - add sleep to keep the test happy.

labkey-bpatel committed:
Revert commented out setup code done for local testing.

labkey-bpatel committed:
Update Investigator facing queries.

labkey-jeckels reviewed (approved):
Looks OK from the dev side, but it looks like there's still the open question on the test timeout.

labkey-bpatel committed:
Uncomment ViewBillingQueries test -- trying WAIT_FOR_JAVASCRIPT & sleep combination.

labkey-bpatel committed:
Increasing test timeout.

labkey-bpatel committed:
Bug fix - Handle insert Row case when user selects from a pre-populated dropdown, or when rowIds are used for Category instead of name.

labkey-bpatel committed:
- Verify viewBillingQueries test differently to make it work with teamcity firefox version.
- Cleanup for groupCategoryAssociations on container delete.
- Fix typo.

labkey-bpatel committed:
Require userid at minimum.

labkey-bpatel committed:
Revert last change.

labkey-bpatel commented:
All the billing tests pass, BVT and DRT doesnt show any billing related failures. Merging..
